### PR TITLE
Added jenkins manager rewritten to not use spring

### PR DESCRIPTION
**Pull request by a-breid-ericsson**
This is the jenkins manager moved from EI. It has been rewritten to work without spring.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:

### CODE REVIEW

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        ResponseEntity response = httpRequest.performRequest();
+        isDeleted = response.getStatusCode() == HttpStatus.SC_MOVED_TEMPORARILY;
+
+        if(!isDeleted) {
+            LOGGER.error("Failed to delete jenkins job. Status code: " + response.getStatusCodeValue() + ". Reason: " + response.getBody());
+        }
+
+        return isDeleted;
+    }
+
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Add java-doc, we said we should add it to every method, except if it's super obvious :)

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    private String getCrumb() throws URISyntaxException, JSONException {
+        String crumb = "";
+        HttpRequest httpRequest = new HttpRequest(HttpMethod.GET);
+
+        httpRequest.setJenkinsBaseUrl(jenkinsBaseUrl).addHeader("Authorization", "Basic " + encoding)
+                .addHeader("Content-type", MediaType.APPLICATION_JSON).setEndpoint("/crumbIssuer/api/json");
+
+        ResponseEntity response = httpRequest.performRequest();
+        System.out.println("HOW == " + response);
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Remove´?

File: src/main/java/utils/HttpRequest.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    @Getter
+    @Setter
+    protected String jenkinsBaseUrl;
+
+    @Getter
+    @Setter
+    protected String endpoint;
+    @Getter
+    protected Map<String, String> params;
+
```
Comment (e-edling-ericsson) [inline src/main/java/utils/HttpRequest.java]: Remove one line

File: src/main/java/utils/HttpRequest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+     * Function that set the body of the http request.
+     *
+     * @param body , the file with body content to be set in the http request.
+     */
+    public void setBody(File file) {
+        String fileContent = "";
+        try {
+            fileContent = FileUtils.readFileToString(file, "UTF-8");
+        } catch (IOException e) {
+            LOGGER.error("Failed to read the Request body file:" + file.getPath() + ". Message: " + e.getMessage(), e);
```
Comment (e-edling-ericsson) [inline src/main/java/utils/HttpRequest.java]: Should this maybe throw an error? It feels like if this goes wrong, the code should not continue?

File: src/main/java/utils/HttpExecutor.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     * Close existing HttpClient and create a new one.
+     *
+     * @throws IOException
+     */
+    public void recreateHttpClient() {
+        if (this.client != null) {
+            try {
+                this.client.close();
+            } catch (IOException e) {
+                LOGGER.error("Error: {}", e.getMessage());
```
Comment (e-edling-ericsson) [inline src/main/java/utils/HttpExecutor.java]: Print stacktrace as well

File: src/main/java/utils/HttpExecutor.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+     *      containing the json content of the http response and status code from request
+     * */
+    public ResponseEntity executeRequest(HttpRequestBase request) {
+        HttpResponse httpResponse = null;
+
+        try {
+            httpResponse = client.execute(request);
+            return new ResponseEntity(httpResponse);
+        } catch(IOException e) {
+            LOGGER.error(e.getMessage(), e);
```
Comment (e-edling-ericsson) [inline src/main/java/utils/HttpExecutor.java]: Log stacktrace as well

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        boolean success = response.getStatusCode() == HttpStatus.SC_OK;
+
+        if (success) {
+            JSONObject jsonObj = new JSONObject(response.getBody());
+            crumb = jsonObj.getString("crumb");
+        } else {
+            LOGGER.error("Failed to get jenkins crumb. Status code: " + response.getStatusCodeValue() + ". Reason: " + response.getBody());
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Maybe this should also throw an error if it did not succeed? Then the user may or may not handle it depending on their implementation

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+        httpRequest.setJenkinsBaseUrl(jenkinsBaseUrl).addHeader("Authorization", "Basic " + encoding)
+                .addHeader("Content-type", MediaType.APPLICATION_JSON).addHeader("Jenkins-Crumb", crumb)
+                .setEndpoint(endpoint);
+
+        ResponseEntity response = httpRequest.performRequest();
+        isDeleted = response.getStatusCode() == HttpStatus.SC_MOVED_TEMPORARILY;
+
+        if(!isDeleted) {
+            LOGGER.error("Failed to delete jenkins job. Status code: " + response.getStatusCodeValue() + ". Reason: " + response.getBody());
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Maybe this should also throw an error if it did not succeed? Then the user may or may not handle it depending on their implementation

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+        httpRequest.setJenkinsBaseUrl(jenkinsBaseUrl).addHeader("Authorization", "Basic " + encoding)
+                .addHeader("Content-type", MediaType.APPLICATION_JSON)
+                .setEndpoint(endpoint);
+
+        ResponseEntity response = httpRequest.performRequest();
+        isTriggered = response.getStatusCode() == HttpStatus.SC_OK;
+
+        if(!isTriggered) {
+            LOGGER.error("Failed to check if job was triggered. Status code: " + response.getStatusCodeValue() + ". Reason: " + response.getBody());
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Maybe this should also throw an error if it did not succeed? Then the user may or may not handle it depending on their implementation

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+        httpRequest.setJenkinsBaseUrl(jenkinsBaseUrl).addHeader("Authorization", "Basic " + encoding)
+                .addHeader("Content-type", MediaType.APPLICATION_JSON).addParam("token", job_token)
+                .setEndpoint(endpoint);
+
+        ResponseEntity response = httpRequest.performRequest();
+        success = response.getStatusCode() == HttpStatus.SC_CREATED;
+
+        if(!success) {
+            LOGGER.error("Failed to trigger a jenkins job. Status code: " + response.getStatusCodeValue() + ". Reason: " + response.getBody());
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Maybe this should also throw an error if it did not succeed? Then the user may or may not handle it depending on their implementation

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+        httpRequest.setJenkinsBaseUrl(jenkinsBaseUrl).addHeader("Authorization", "Basic " + encoding)
+                .addHeader("Content-type", MediaType.APPLICATION_XML).addHeader("Jenkins-Crumb", crumb)
+                .addParam("name", jobName).setBody(jobXmlData).setEndpoint("/createItem");
+
+        ResponseEntity response = httpRequest.performRequest();
+        success = response.getStatusCode() == HttpStatus.SC_OK;
+
+        if(!success) {
+            LOGGER.error("Failed to create a jenkins job. Status code: " + response.getStatusCodeValue() + ". Reason: " + response.getBody());
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: Maybe this should also throw an error if it did not succeed? Then the user may or may not handle it depending on their implementation

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+    /**
+     * Creates a jenkins job with a given name using the XML data as input for
+     * job configuration
+     *
+     * @param jobName
+     * @param jobXmlData
+     * @return
+     * @throws URISyntaxException
+     * @throws ClientProtocolException
```
Comment (evasiba-ericsson) [inline src/main/java/com/ericsson/esj/JenkinsManager.java]: It does not look like it throws all these exception

File: src/test/java/com/ericsson/esj/jenkins_manager_test/JenkinsManagerExecutionTest.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import com.ericsson.esj.JenkinsManager;
+
+public class JenkinsManagerExecutionTest {
+
+    private static final String username = "admin";
```
Comment (evasiba-ericsson) [inline src/test/java/com/ericsson/esj/jenkins_manager_test/JenkinsManagerExecutionTest.java]: These values should be set in pom. And it should be possible to be provided from command line when running maven.

Comment (a-breid-ericsson) [reply src/test/java/com/ericsson/esj/jenkins_manager_test/JenkinsManagerExecutionTest.java]: The values should be given from a test, test properties. And there will be tasks to implement test in the future.

File: src/main/java/com/ericsson/esj/JenkinsManager.java (hunk 12 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/esj/JenkinsManager.java]: It does now :)

File: src/main/java/com/ericsson/esj/JenkinsManager.java [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+    public boolean deleteJob(String jobName) throws URISyntaxException {
+        boolean isDeleted = false;
+        HttpRequest httpRequest = new HttpRequest(HttpMethod.POST);
+        if (jobName == null) {
+            jobName = DEFAULT_JOB_NAME;
+        }
+
+        String endpoint = "/job/" + jobName + "/doDelete";
+
+        httpRequest.setJenkinsBaseUrl(jenkinsBaseUrl).addHeader("Authorization", "Basic " + encoding)
```
[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/esj/JenkinsManager.java]: Maybe improved?

File: src/main/java/com/ericsson/esj/helpers/JenkinsXmlData.java [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+     *
+     * @return
+     */
+    public String getXmlAsString() {
+        xmlJsonData.getJSONObject("project").put("builders", builders);
+        xmlJsonData.getJSONObject("project").put("properties", properties);
+        System.out.println(xmlJsonData.toString(2));
+        String xmlDataString = XML.toString(xmlJsonData);
+        xmlDataString = removeExtraGroovyParams(xmlDataString);
+        // System.out.println(xmlDataString);
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/helpers/JenkinsXmlData.java]: Remove this

File: src/main/java/com/ericsson/esj/helpers/JenkinsXmlData.java [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+     * @param parametertypeKey
+     */
+    private void validatePropertiesObject(String parametertypeKey) {
+        boolean hasHudsonParametersDefinitionKey = properties.has(HUDSON_PARAMETERS_DEFENITION_KEY);
+        if (!hasHudsonParametersDefinitionKey) {
+            JSONObject parametersDefinitionProperty = new JSONObject();
+            properties.put(HUDSON_PARAMETERS_DEFENITION_KEY, parametersDefinitionProperty);
+        }
+
+        boolean hasParameterDefinitions = properties.getJSONObject(HUDSON_PARAMETERS_DEFENITION_KEY)
```
Comment (e-edling-ericsson) [inline src/main/java/com/ericsson/esj/helpers/JenkinsXmlData.java]: HUDSON_PARAMETERS_DEFENITION_KEY ->HUDSON_PARAMETERS_DEFINITION_KEY.  Spelling mistake, same with the other constants :)

File: src/test/java/com/ericsson/esj/jenkins_manager_test/JenkinsManagerExecutionTest.java [hunk 17]
```diff
@@ -22,40 +38,100 @@
             Arrays.asList("Test_Job1", "Test_Job2", "Test_Job3", "Test_Job4", "Test_Job5", "Test_Job6", "Test_Job7",
                     "Test_Job8", "Test_Job9", "Test_Job10", "Test_Job11", "Test_Job12", "Test_Job13", "Test_Job14",
                     "Test_Job15", "Test_Job16", "Test_Job17", "Test_Job18", "Test_Job19", "Test_Job20"));
+    static JenkinsManager jenkinsManager;
 
-    public static void main(String[] args) throws URISyntaxException, IOException, InterruptedException {
```
Comment (e-edling-ericsson) [inline src/test/java/com/ericsson/esj/jenkins_manager_test/JenkinsManagerExecutionTest.java]: Maybe use junit or something like that for the tests instead? Feels better to use some common framework

**Comment by l-ohlen-ericsson** [comment]
Why **java** as a part of the repo name? It will limit what is shared (could besides java be scripts, javascripts etc). The name is somewhat clashing (except the padding with -java) an other library, this would be confusing. What about **eiffel-commons** (Apache style naming convention)?

File: src/main/java/com/ericsson/esj/helpers/JenkinsXmlData.java (hunk 16 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/esj/helpers/JenkinsXmlData.java]: Done

**Pull request merged**